src/platform/packages/shared/kbn-esql-utils/src/utils/cascaded_documents_helpers.ts

```diff
@@ -0,0 +1,489 @@
+        // if there's a where command targeting the group in this current iteration,
+        // then this specific query can only be grouped by the current group, pivoting on any other columns though they exist
+        // in the query would be invalid, hence we clear out any previously added group by fields since they are no longer valid
+        groupByFields.splice(0, groupByFields.length);
+      }
+
+      // add the current group and break out of the loop
+      // since there's no need to continue processing other groups
+      // as they are not valid in this context
+      groupByFields.push({
+        field: groupFieldName,
+        type:
+          groupFieldNode.definition.type === 'function'
+            ? groupFieldNode.definition.name
+            : groupFieldNode.definition.type,
+      });
+
+      break;
+    }
+
+    if (isFunctionExpression(groupFieldNode.definition)) {
+      const functionName = groupFieldNode.definition.name;
+      if (!isSupportedStatsFunction(functionName)) {
+        continue;
+      }
+    }
+
+    groupByFields.push({
+      field: groupFieldName,
+      type:
+        groupFieldNode.definition.type === 'function'
+          ? groupFieldNode.definition.name
+          : groupFieldNode.definition.type,
+    });
+  }
+
+  Object.values(summarizedStatsCommand.aggregates).forEach((aggregate) => {
+    appliedFunctions.push({
+      identifier: removeBackticks(aggregate.field), // we remove backticks to have a clean identifier that gets displayed in the UI
+      aggregation:
+        (aggregate.definition as ESQLFunction).operator?.name ?? aggregate.definition.text,
+    });
+  });
+
+  return { groupByFields, appliedFunctions };
+};
+
+export interface CascadeQueryArgs {
+  /**
+   * anchor query for generating the next valid query
+   */
+  query: AggregateQuery;
+  /**
+   * Node type (group or leaf) for which we are constructing the cascade query
+   */
+  nodeType: NodeType;
+  /**
+   * Node path for the current node in the cascade experience we'd like to generate a query for
+   */
+  nodePath: string[];
+  /**
+   * Mapping of node paths to their corresponding values, used to populate the query with literal values
+   */
+  nodePathMap: Record<string, string>;
+}
+
+/**
+ * Constructs a cascade query from the provided query, based on the node type, node path and node path map.
+ */
+export const constructCascadeQuery = ({
+  query,
+  nodeType,
+  nodePath,
+  nodePathMap,
+}: // @ts-expect-error -- temporary stop gap to fix deployment
+CascadeQueryArgs): AggregateQuery => {
+  const EditorESQLQuery = EsqlQuery.fromSrc(query.esql);
+
+  const dataSourceCommand = mutate.generic.commands.find(
+    EditorESQLQuery.ast,
+    (cmd) => cmd.name === 'from'
+  ) as ESQLCommand<'from'> | undefined;
+
+  const queryRuntimeFields = Array.from(mutate.commands.stats.summarize(EditorESQLQuery)).map(
+    (command) => command.newFields
+  );
+
+  if (!dataSourceCommand) {
```

Comment:
Move this before the queryRuntimeFields as if there is no source command it should break immediately